Introduce Typed objects before referring to them (#121)
apple/pkl: docs/modules/language-tutorial/pages/02_filling_out_a_template.adoc

}
----

[NOTE]
====

> So far, we have only worked with link:{uri-stdlib-Dynamic}[Dynamic] objects. The other type available to us is link:{uri-stdlib-Typed}[Typed] objects. _Amending_ allows us to override, amend, and add new properties to a dynamic object, but a typed object will only let us amend or override existing properties, not add entirely new ones.

Reviewer: One code style note: We keep to one sentence per line (gives somewhat nicer diffing over time).

For presentation purposes, we prefer the _second_ person (it's more activating).

Although I agree with this fix, I feel some wordsmithing is called for, especially because `Listing` and `Mapping` live at the same level as `Dynamic` and `Typed`. Leaving it to you to accept/alter:
```suggestion
So far, you have only worked with link:{uri-stdlib-Dynamic}[Dynamic] objects.
Pkl also offers link:{uri-stdlib-Typed}[Typed] objects.
_Amending_ allows you to override, amend, and add new properties to a dynamic object.
Typed objects will only let you amend or override existing properties, not add entirely new ones.
```

Author: Thanks, that's an improvement; applied 🙏🏻

The xref:03_writing_a_template.adoc[next part of the tutorial] discusses types in more detail.

> There, you see that amending _never changes the type_ of a typed object.

Reviewer: This one, I'd revert, because `Dynamic` is also a type that doesn't change by amending.
```suggestion
There, you see that amending _never changes the type_ of the object.
```

Author: Ah, got it! I somehow was under the impression that there was some type associated with a `Dynamic` object, too, based on its properties, elements, and entries, but it seems to be more akin to the Lua table data type then.
